src/main/java/uk/gov/hmcts/reform/juddata/camel/util/JrdUserProfileUtil.java

```diff
@@ -0,0 +1,271 @@
+        filteredMap.forEach((k, v) -> filteredList.addAll(v));
+
+        return List.copyOf(filteredList);
+    }
+
+    /**
+     * Iterate through the list of user profiles and group them by object id,
+     * remove the entries where the object id is correctly associated with no more than 1 distinct personal code,
+     * respecting 1-1 mapping between object ids and personal codes.
+     * @param userProfiles original list of user profiles
+     * @return map containing only the object ids where the associated personal codes are invalid
+     */
+    private Map<String, List<JudicialUserProfile>> filterAndGroupByObjectId(
+            List<JudicialUserProfile> userProfiles) {
+
+        return userProfiles.stream()
+                .filter(userProfile -> StringUtils.isNotBlank(userProfile.getObjectId()))
+                .collect(collectingAndThen(groupingBy(JudicialUserProfile::getObjectId), map -> {
+                    map.values().removeIf(l -> l.stream()
+                            .map(JudicialUserProfile::getPersonalCode)
+                            .distinct().count() < 2);
+                    return map; }));
+    }
+
+
+    /**
+     * Iterate through the list of user profiles and group them by personal code,
+     * remove the entries where the personal code is correctly associated with no more than 1 distinct object id,
+     * respecting 1-1 mapping between personal codes and object ids.
+     * @param userProfiles original list of user profiles
+     * @return map containing only the personal codes where the associated object ids are invalid
+     */
+    private Map<String, List<JudicialUserProfile>> filterAndGroupByPersonalCode(
+            List<JudicialUserProfile> userProfiles) {
+
+        return userProfiles.stream()
+                .filter(userProfile -> StringUtils.isNotBlank(userProfile.getObjectId()))
+                .collect(collectingAndThen(groupingBy(JudicialUserProfile::getPersonalCode), map -> {
+                    map.values().removeIf(l -> l.stream()
+                            .map(JudicialUserProfile::getObjectId)
+                            .distinct().count() < 2);
+                    return map; }));
+    }
+
+    private List<JudicialUserProfile> getInvalidRecords(List<JudicialUserProfile> userProfiles) {
+        //get list of user profiles with invalid object id
+        var userProfilesWithInvalidObjectIds = filterByObjectId(userProfiles);
+
+        //get list of user profiles with invalid personal code
+        var userProfilesWithInvalidPersonalCodes = filterByPersonalCode(userProfiles);
+
+        var invalidRecordsList = new ArrayList<JudicialUserProfile>();
+        Stream.of(userProfilesWithInvalidObjectIds, userProfilesWithInvalidPersonalCodes)
+                .forEach(invalidRecordsList::addAll);
+
+        return List.copyOf(invalidRecordsList);
+    }
+
+    public void remove(List<JudicialUserProfile> userProfilesToBeDeleted, List<JudicialUserProfile> userProfiles) {
```

Comment:
I think this is being called only in this class - can it be private?

Reply:
The methods have been made public for unit testing purposes. But I can make them private if that's an issue.